felix/bpf/conntrack/bpf_scanner.go

```diff
@@ -0,0 +1,305 @@
+	}, []string{"type"})
+	counterVecConntrackEntriesDeleted = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "felix_bpf_conntrack_entries_deleted",
+		Help: "Cumulative number of entries deleted from the conntrack table, grouped by type.",
+	}, []string{"type"})
+	summaryCleanerExecTime = prometheus.NewSummary(prometheus.SummaryOpts{
+		Name: "felix_bpf_conntrack_cleaner_seconds",
+		Help: "Time taken to run the conntrack cleaner BPF program.",
+	})
+)
+
+func registerConntrackMetrics() {
+	registerOnce.Do(func() {
+		prometheus.MustRegister(
+			gaugeVecConntrackEntries,
+			counterVecConntrackEntriesDeleted,
+			summaryCleanerExecTime,
+		)
+	})
+}
+
+// BPFProgLivenessScanner is a scanner that uses a BPF program to scan the
+// conntrack table for expired entries.  The BPF program does the entry
+// deletion, taking care to delete forward and reverse NAT entries together,
+// thus minimising the window where only one entry is present.
+//
+// Note: the tests for this object are largely in the bpf/ut package, since
+// we require a privileged environment to test the BPF program.
+type BPFProgLivenessScanner struct {
+	ipVersion int
+	timeouts  Timeouts
+	logLevel  BPFLogLevel
+
+	bpfExpiryProgram *libbpf.Obj
+}
+
+func NewBPFProgLivenessScanner(
+	ipVersion int,
+	timeouts Timeouts,
+	bpfLogLevel BPFLogLevel,
+) (*BPFProgLivenessScanner, error) {
+	if ipVersion != 4 && ipVersion != 6 {
+		return nil, fmt.Errorf("invalid IP version: %d", ipVersion)
+	}
+	if bpfLogLevel != BPFLogLevelDebug && bpfLogLevel != BPFLogLevelNone {
+		return nil, fmt.Errorf("invalid BPF log level: %s", bpfLogLevel)
+	}
+	s := &BPFProgLivenessScanner{
+		ipVersion: ipVersion,
+		timeouts:  timeouts,
+		logLevel:  bpfLogLevel,
+	}
+	_, err := s.ensureBPFExpiryProgram()
+	if err != nil {
+		return nil, err
+	}
+	registerConntrackMetrics()
+	return s, nil
+}
+
+func (s *BPFProgLivenessScanner) ensureBPFExpiryProgram() (*libbpf.Obj, error) {
+	if s.bpfExpiryProgram != nil {
+		return s.bpfExpiryProgram, nil
+	}
+
+	binaryToLoad := path.Join(bpfdefs.ObjectDir,
+		fmt.Sprintf("conntrack_cleanup_%s_v%d.o", s.logLevel, s.ipVersion))
+	obj, err := libbpf.OpenObject(binaryToLoad)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load conntrack cleanup BPF program: %w", err)
+	}
+
+	success := false
+	defer func() {
+		if !success {
+			err := obj.Close()
+			if err != nil {
+				log.WithError(err).Error("Error closing BPF object.")
+			}
+		}
+	}()
+
+	ctMapParams := MapParams
+	if s.ipVersion == 6 {
+		ctMapParams = MapParamsV6
+	}
+	configuredGlobals := false
+	pinnedCTMap := false
+	for m, err := obj.FirstMap(); m != nil && err == nil; m, err = m.NextMap() {
+		// In case of global variables, libbpf creates an internal map <prog_name>.rodata
+		// The values are read only for the BPF programs, but can be set to a value from
+		// userspace before the program is loaded.
+		mapName := m.Name()
+		if m.IsMapInternal() {
+			if strings.HasPrefix(mapName, ".rodata") {
```

Comment:
I noticed recently that there was a change in naming and `Contains` is now required